### DISCUSSION TITLE
Manually using rsync in piped mode (--server etc)

### DISCUSSION THREAD

**Discussion by psy0rz** [RsyncProject/rsync#278] [Q&A]
I want to manually connect the pipes from an rsync --sender to a rsync client for my project.

The reasons are: The project should support two remotes at the same time, and i already manage my own ssh traffic and debug logging. 

So instead of trying to recreate a rsync command that connects to the correct remote, i want it to use my own pipes.

I tested with something like:
```
mkfifo /tmp/f
rsync --server --sender -avx root@1.1.1.1:/tmp/a /tmp/b </tmp/f | rsync --server -avx /tmp/a /tmp/b  > /tmp/f
```

which of course doesnt work.

Is it something as easy as: connect stdout of the target to the stdin of the server/sender? (I guess not since the verbose option also sends to std out.)

Or is it much more convoluted and a bad idea?

**Comment by WayneD** [comment]
Firstly, a --server doesn't support remote args: the server **is** remote and has to be run on the remote host. Secondly, the protocol doesn't support such an idea as 2 servers, since the server & client each do slightly different things at the start of the transfer. Also, it is only the client that outputs what is going on and to read in things like --exclude-from data, so there must be a client for things to work right.

What you might be looking for is a daemon, which is a way for a client rsync to talk to a remote rsync via a socket (which can be already running).  You could tunnel a socket via the ssh protocol and connect via normal daemon syntax to the local port (that is being handled by ssh) or you could use something akin to the `rsync-ssl` script to let rsync run a program to get a pipe to talk to a daemon and replace the remote shell command with something that hands over a stdin/stdout to whatever remote daemon you want to access (perhaps running it, perhaps not).  In all cases, you'd need to be running a client rsync command to initiate things, though.

**Comment by psy0rz** [comment]
I will have a look at rsync-ssl.

So normally when usng ssh, on the server side it will start an rsync --server --sender.

What about the client side? Can i start that manually as well? Or isnt that possible since it will normally determine this internally. (e.g. it starts the --server remote and becomes the client itself)

Running in daemon mode is convoluted and complicates things even more, so wont do that. (then its a better option to use rsync like its used normally with ssh and just not support push/pull mode)
